docs: Add gRPC proxy chapter (#736)
oasisprotocol/docs: docs/node/grpc.md

---
description: gRPC proxy for Oasis node
---

# gRPC proxy for your Oasis node

> The Oasis node API is exposed via the [gRPC interface]. It enables communication

Reviewer: ```suggestion
The Oasis node API is exposed via the [gRPC protocol]. It enables communication
```

with external applications such as the Oasis CLI, dApps running in your browser
and need to perform actions on the consensus layer or on a non-EVM

> ParaTime, monitoring and controlling the node and similar.

Reviewer: ```suggestion
that need to perform actions on the consensus layer or a ParaTime,
monitoring and controlling the node and similar.
```

:::tip Web3 gateway

> The Oasis gRPC **is not related to the standardized Web3 interface**. For

Reviewer: ```suggestion
The Oasis gRPC **is not related to the standardized Web3 JSON-RPC API**. For
```

> EVM-compatible dApps we suggest that you set up a [Web3 gateway] instead and

Reviewer: ```suggestion
EVM-compatible dApps we suggest that you set up a [Web3 gateway] instead
```

which is also compatible with other Ethereum tooling.

:::

> The gRPC proxy performs the following:

Reviewer: ```suggestion
The gRPC proxy may perform the following:
```

- transforms the gRPC into HTTP methods so that you can access the Oasis node

>   directly from your browser,

Reviewer: ```suggestion
- makes gRPC available to in-browser applications via gRPC-Web,
```

- filters out control methods such as shutting down your node,
- authentication,
- load balances the traffic to multiple Oasis nodes.

This chapter will show you how to set up a **public** gRPC endpoint using Envoy
so that it exposes only a **safe subset of the [Oasis RPC services]**. The
final section presents an alternative approach which **securely tunnels a Unix
socket over the network**, so it can safely be used by the client, but **does
not filter out any services**.

> :::danger Never expose RPC Unix socket directly!

Reviewer: ```suggestion
:::danger Never expose the UNIX socket directly!
```

The `oasis-node` deliberately exposeses the RPC interface only via an AF_LOCAL
socket called `internal.sock` located in the node's data directory.

> The socket should **never be directly exposed over the network** as it has no

Reviewer: ```suggestion
This socket should **never be directly exposed over the network** as it has no
```

authentication and allows full control—including shutting down your node.

:::

[gRPC interface]: https://github.com/oasisprotocol/oasis-core/blob/master/docs/oasis-node/rpc.md
[Web3 gateway]: web3.md
[Oasis RPC services]: https://github.com/oasisprotocol/oasis-core/blob/master/docs/oasis-node/rpc.md#services

## gRPC Proxy with Envoy {#envoy}

Let's set up a typical public Oasis endpoint using the [Envoy HTTP proxy] with
the following behavior:

- whitelisted methods for checking the network status, performing queries and
  submitting transactions and no control methods and queries that are CPU and
  I/O intensive,

> - lives on `grpc.my-oasis-node.com` with TLS-enabled connection and Let's
>   encrypt certificates.

Reviewer 2: ```suggestion
- lives on `grpc.my-oasis-node.com` with TLS-enabled connection and
  certificates that you already have from Let's Encrypt.
```

trying to clarify that this envoy config doesn't set up certbot

# Envoy gRPC-web proxy configuration
---
admin:
  address:
    socket_address:
      address: "127.0.0.1"
      port_value: 10000
static_resources:
  listeners:
    - name: listener_0
      address:
        socket_address:
          address: "0.0.0.0"
          port_value: 443
      filter_chains:
        - filters:
            - name: envoy.filters.network.http_connection_manager
              typed_config:
                "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
                codec_type: AUTO
                stat_prefix: ingress_http
                access_log:
                  - name: envoy.file_access_log
                    typed_config:
                      "@type": type.googleapis.com/envoy.extensions.access_loggers.file.v3.FileAccessLog
                      path: /dev/stdout
                route_config:
                  virtual_hosts:
                    - name: vh_0
                      domains:
                        - "*"
                      routes:
                        - match:
                            safe_regex:
                              regex: "/oasis-core\\.(\

>     Beacon/(\

Reviewer 2: oh this might not work when it's spaced out like this. we have it in this format in our internal repo, but with an additional filter later to remove spaces

Author: Using `"` will trim leading whitespaces of every line and it will replace the new lines with a single whitespace (which we do not want). Then, using `\` at the end of each new line will make the newlines not to be replaced with a whitespace.

Ref: https://yaml.org/spec/1.2.2/#731-double-quoted-style

Tested it out here: https://eemeli.org/yaml-playground/

Reviewer 2: oo nice find

---
description: gRPC proxy for Oasis node
---

# gRPC proxy for your Oasis node

> The Oasis node uses a [gRPC interface] which enables external applications such

Reviewer: ```suggestion
The Oasis node uses the [gRPC protocol] which enables external applications such
```

as the Oasis CLI, cross-layer and Oasis Wasm dApps, native wallets, indexers,
system monitors, health checks and similar to access various consensus and
ParaTime services.

> :::info

Reviewer: Should this be caution or danger?

The RPC interface is ONLY exposed via an AF_LOCAL socket
called `internal.sock` located in the node's data directory.
**This interface should NEVER be directly exposed over the network as it has no
authentication and allows full control, including shutdown, of a node.**

:::

This chapter shows you how to **tunnel a Unix socket over the network**, so it
can safely be used on your client, or set up a **public** gRPC endpoint using
Envoy or NginX and expose only a **safe subset of the [Oasis RPC services]**.

> ## Tunnel Unix socket via SSH

Reviewer: I would put this after the more production options.

SSH supports forwarding a Unix socket over a secure layer. The command below
will establish a secure shell to the `my-oasis-node.com` server and then tunnel
the `internal.sock` file inside the data directory to a Unix socket inside your
home folder:

```shell
ssh oasis@my-oasis-node.com -L /home/user/oasis-node-internal.sock:/node/data/internal.sock
```

The `/home/user/oasis-node-internal.sock` can now be used to safely connect to
the Oasis node **as if it was running locally without filtering any services**.
For example, using the [Oasis CLI]:

```shell
oasis network add-local my-oasis-node unix:/home/user/oasis-node-internal.sock
```

```shell
oasis network status --network my-oasis-node
```

:::tip Permanent SSH channel

To make a tunneled Unix socket over SSH permanent, consider using [autossh].

:::

[Oasis CLI]: https://github.com/oasisprotocol/cli/blob/master/docs/network.md#add-local
[autossh]: https://www.harding.motd.ca/autossh/

## Set up gRPC Proxy with Envoy {#envoy}

> ## Set up gRPC Proxy with Nginx {#nginx}

Reviewer: Maybe also [Caddy](https://caddyserver.com/).

Author: Let's do this in a separate PR. I also removed Nginx from the chapter.
